File: queue.rb
require "pry"

class Queue
  def initialize
    @list = []
    @tail = 0
  end

  def enqueue(v)
    @list[@tail] = v
    @tail += 1
  end

  def dequeue
    v = @list.shift()
    @tail -= 1
    v
  end

  def show
    @list.to_s
  end
end

File: queue.js
//Example of how a queue data structure works using js classes.
class Queue {
    constructor() {
        this.list = [];
        this.head = 0;
        this.tail = 0;
    }

    enqueue(v) {
        this.list[this.tail] = v;
        this.tail++;
    }

    dequeue() {
        const element = this.list[this.head];
        this.list.shift();
        if (!element) return;
        this.head++;
        return element;
    }

    size() {
        return this.tail - this.head;
    }

    peek() {
        return this.list[0];
    }

    show() {
        return this.list;
    }
    isEmpty(){
        return this.size() === 0;
    }
}

class QueueLinkedList {
    constructor() {
        this.list = new LinkedList();
    }

    enqueue(v) {
        this.list.addToTail(v);
    }

    dequeue() {
        return this.list.removeHead();
    }

    size() {
        return this.list.size();
    }
}
